File: 1_sorting_algorithms/sorting_algorithms/go/algos/countingsort.go
package algos

func CountingSort(arr []int) {
	if len(arr)==0 {
		return
	}
	low:=arr[0]
	high:=arr[0]
	n:=len(arr)
	for i:=range n {
		if arr[i]<low {
			low = arr[i]
		} else if arr[i]>high {
			high=arr[i]
		}
	}

	// make array start from 0
	for i:=range n {
		arr[i]-=low
	}
	k:=high-low

	// init counts count array
	counts:=make([]int,k+1)
	for i:=range n {
		counts[arr[i]]+=1
	}

	// cummulate the sums to index correctly
	for i:=1;i<k+1;i++ {
		counts[i]+=counts[i-1]
	}

	// store output in reverse order and store into original array
	output:=make([]int,n)
	for i:=n-1;i>=0;i-- {
		output[counts[arr[i]]-1]=arr[i]
		counts[arr[i]]-=1
	}
	copy(arr,output)

	// fix the array from the start
	for i:=range n {
		arr[i]+=low
	}
}

File: 1_sorting_algorithms/sorting_algorithms/rs/src/algos/countingsort.rs
pub fn countingsort(arr: &mut Vec<i32>) {
    // init aux
    let (mut l, mut r) = (arr[0], arr[0]);
    for i in 1..arr.len() {
        if arr[i] < l {
            l = arr[i];
        } else if arr[i] > r {
            r = arr[i];
        }
    }
    let k = r - l;
    let mut aux = vec![0; k as usize + 1];
    for num in &*arr {
        aux[(*num - l) as usize] += 1;
    }
    // dump into arr
    let mut i = 0;
    for num in 0..k + 1 {
        while aux[num as usize] > 0 {
            arr[i] = num + l;
            i += 1;
            aux[num as usize] -= 1;
        }
    }
}
